### ISSUE TITLE
Optimize memory usage for large projects

### ISSUE THREAD

**Issue by AaronFriel** [uber-go/nilaway#127]
I've attempted to run `nilaway` on Pulumi's projects, including our core repository, but the memory usage is substantial, it looks like around 9GiB resident to scan `github.com/pulumi/pulumi/pkg/cmd/...` (the packages at "cmd" and below), and running it on the entire repo results in the OOM killer even though the VM had access to about 20GiB of memory.

Do you have any guidance related to memory requirements, and how we could run this against large projects of ours? 

Edit:

It looks like [there is configuration to include or exclude packages](https://github.com/uber-go/nilaway/wiki/Configuration), but in terms of excluding large offenders, it's hard to know which might have the most impact on scanning. Is it possible to measure the space cost of packages?

**Comment by Equationzhao** [comment]
I was trying to scan our project (about 1.2M lines go code) and `nilaway` took all my memory (6G mem and 10G+ swap)

**Comment by yuxincs** [comment]
Hey! We indeed have some inefficiencies in NilAway that we have identified (our [trigger merging logic](https://github.com/uber-go/nilaway/blob/a267567c6ffff900df0c3394d031ee70079ec8df/annotation/full_trigger.go#L190), and our [backprop algorithm](https://github.com/uber-go/nilaway/blob/main/assertion/function/assertiontree/backprop.go#L752), in case anyone is curious), and we are brewing some PRs to optimize those.

It was not a huge problem for us since we were internally leveraging the incremental build system (i.e., bazel and nogo), such that packages are built incrementally, and facts are serialized to disk and read on demand.

This is a big problem when running NilAway as a standalone linter since the default [`singlechecker`](https://pkg.go.dev/golang.org/x/tools/go/analysis/singlechecker) driver stores everything in memory. 

In the meantime, as you pointed out, it's possible to exclude packages from analysis (usually we exclude 3rd party dependencies). As for getting an estimate of memory requirement, it is usually proportional to the number of objects (defined [here](https://github.com/golang/example/tree/master/gotypes#objects)) in each package.
